resteasy-client-vertx/src/main/java/org/jboss/resteasy/client/jaxrs/engines/vertx/VertxClientHttpEngine.java

```diff
@@ -0,0 +1,283 @@
+package org.jboss.resteasy.client.jaxrs.engines.vertx;
+
+import io.vertx.core.MultiMap;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.http.HttpClientOptions;
+import io.vertx.core.http.HttpClientRequest;
+import io.vertx.core.http.HttpClientResponse;
+import io.vertx.core.http.HttpMethod;
+import io.vertx.core.http.RequestOptions;
+import org.jboss.resteasy.client.jaxrs.engines.AsyncClientHttpEngine;
+import org.jboss.resteasy.client.jaxrs.internal.ClientConfiguration;
+import org.jboss.resteasy.client.jaxrs.internal.ClientInvocation;
+import org.jboss.resteasy.client.jaxrs.internal.ClientResponse;
+
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.SSLContext;
+import javax.ws.rs.ProcessingException;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.client.Invocation;
+import javax.ws.rs.client.InvocationCallback;
+import javax.ws.rs.client.ResponseProcessingException;
+import javax.ws.rs.core.Response;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+import io.vertx.core.http.HttpClient;
+import io.vertx.core.http.HttpHeaders;
+import org.jboss.resteasy.util.CaseInsensitiveMap;
+
+public class VertxClientHttpEngine implements AsyncClientHttpEngine {
+
+    private final Vertx vertx;
+    private final HttpClient httpClient;
+
+    public VertxClientHttpEngine() {
+        this.vertx = Vertx.vertx();
+        this.httpClient = vertx.createHttpClient();
+    }
+
+    public VertxClientHttpEngine(final Vertx vertx, final HttpClientOptions options) {
+        this.vertx = vertx;
+        this.httpClient = vertx.createHttpClient(options);
+    }
+
+    public VertxClientHttpEngine(final Vertx vertx) {
+        this(vertx, new HttpClientOptions());
+    }
+
+    public VertxClientHttpEngine(final HttpClient client) {
+        this.vertx = null;
+        this.httpClient = client;
+    }
+
+    @Override
+    public <T> Future<T> submit(final ClientInvocation request,
+                                final boolean buffered,
+                                final InvocationCallback<T> callback,
+                                final ResultExtractor<T> extractor) {
+        CompletableFuture<T> future = submit(request).thenCompose(response -> {
+            CompletableFuture<T> tmp = new CompletableFuture<>();
+            vertx.executeBlocking(promise -> {
+                try {
+                    T result = extractor.extractResult(response);
+                    tmp.complete(result);
+                } catch (Exception e) {
+                    tmp.completeExceptionally(e);
+                }
+            }, ar -> {
+                //
```

Comment:
Should you complete in this callback, or do you expect blocking downstream stages?

Reply:
we actually complete from the worker thread to avoid blocking the event-loop so we don't care much of the result on the event loop